File: solutions/graphs/695-max-area-of-island.py
"""
# Difficulty: Medium

# 695. Max Area Of Island

You are given an m x n binary matrix grid. An island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.

The area of an island is the number of cells with a value 1 in the island.

Return the maximum area of an island in grid. If there is no island, return 0.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[[0,0,1,0,0,0,0,1,0,0,0,0,0]</dd>
<dt>Output:</dt>
<dd>1</dd>
<dt>Explanation:</dt>
<dd>Maximum island area is 6 square units</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(M × N)
**Space Complexity**: O(M × N)

### INTUITION:
This is similar to "Number of Islands" but instead of counting islands, we need to find the largest island by area. We use DFS/BFS to explore each island and calculate its area, keeping track of the maximum area found.

### APPROACH:
1. **Iterate through grid**: Check each cell
2. **Find land**: When we find a 1 (land), start exploring the island
3. **DFS/BFS exploration**: Count all connected land cells (area)
4. **Track maximum**: Keep track of the largest area found
5. **Mark visited**: Avoid double-counting cells

### WHY THIS WORKS:
- DFS/BFS explores connected components completely
- Each exploration returns the area of that island
- We track the maximum area across all islands
- In-place marking ensures each cell is counted exactly once

### EXAMPLE WALKTHROUGH:
```
Grid: [[0,0,1,0,0,0,0,1,0,0,0,0,0],
       [0,0,0,0,0,0,0,1,1,1,0,0,0],
       [0,1,1,0,1,0,0,0,0,0,0,0,0],
       [0,1,0,0,1,1,0,0,1,0,1,0,0]]

Islands found:
- (0,2): area = 1
- (0,7), (1,7), (1,8), (1,9): area = 4
- (2,1), (2,2), (3,1): area = 3
- (2,4), (3,4), (3,5): area = 3
- (3,8): area = 1
- (3,10): area = 1

Maximum area = 4
```

### TIME COMPLEXITY:
O(M × N)
We visit each cell at most once

### SPACE COMPLEXITY:
O(M × N)
For recursion stack in worst case

### EDGE CASES:
- **All water**: Return 0 (no islands)
- **Single cell island**: Return 1
- **Entire grid is one island**: Return m * n
- **Multiple islands**: Return area of largest
- **Empty grid**: Return 0

</details>
"""

from collections import deque
from typing import Any


class Solution:
    def maxAreaOfIsland(self, grid: list[list[int]]) -> int:
        """
        Find maximum island area using DFS.

        Args:
            grid: 2D binary grid where 1 = land, 0 = water

        Returns:
            Maximum area of any island

        Time Complexity: O(M × N)
        Space Complexity: O(M × N) for recursion stack
        """
        if not grid or not grid[0]:
            return 0

        rows, cols = len(grid), len(grid[0])
        max_area = 0

        def dfs(r: Any, c: Any) -> Any:
            """DFS to calculate area of island starting at (r,c)."""
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == 0:
                return 0

            grid[r][c] = 0  # Mark as visited

            # Count current cell + all connected cells
            return 1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    area = dfs(r, c)
                    max_area = max(max_area, area)

        return max_area

    def maxAreaOfIslandBFS(self, grid: list[list[int]]) -> int:
        """BFS approach to avoid recursion stack issues."""
        if not grid or not grid[0]:
            return 0

        rows, cols = len(grid), len(grid[0])
        max_area = 0
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    area = 0
                    queue = deque([(r, c)])
                    grid[r][c] = 0

                    while queue:
                        curr_r, curr_c = queue.popleft()
                        area += 1

                        for dr, dc in directions:
                            nr, nc = curr_r + dr, curr_c + dc
                            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 1:
                                grid[nr][nc] = 0
                                queue.append((nr, nc))

                    max_area = max(max_area, area)

        return max_area


def test_solution() -> None:
    """Test cases for Problem 695."""
    solution = Solution()

    # Test case 1: Example grid
    grid1 = [
        [0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
        [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0],
    ]
    result1 = solution.maxAreaOfIslandBFS([row[:] for row in grid1])
    expected1 = 6
    assert result1 == expected1, f"Expected {expected1}, got {result1}"

    # Test case 2: No islands
    grid2 = [[0, 0, 0, 0, 0, 0, 0, 0]]
    result2 = solution.maxAreaOfIslandBFS([row[:] for row in grid2])
    expected2 = 0
    assert result2 == expected2, f"Expected {expected2}, got {result2}"

    # Test case 3: Single cell
    grid3 = [[1]]
    result3 = solution.maxAreaOfIslandBFS([row[:] for row in grid3])
    expected3 = 1
    assert result3 == expected3, f"Expected {expected3}, got {result3}"

    # Test case 4: Multiple islands
    grid4 = [[1, 1, 0, 0, 0], [1, 1, 0, 0, 0], [0, 0, 0, 1, 1], [0, 0, 0, 1, 1]]
    result4 = solution.maxAreaOfIslandBFS([row[:] for row in grid4])
    expected4 = 4
    assert result4 == expected4, f"Expected {expected4}, got {result4}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Quick example
    solution = Solution()
    grid = [[1, 1, 0], [0, 1, 0], [0, 0, 1]]
    print(f"Grid: {grid}")
    # Use BFS to preserve original grid
    print(f"Max area: {solution.maxAreaOfIslandBFS([row[:] for row in grid])}")

File: solutions/graphs/695-max-area-of-island.js
I'll help you create a solution for the Max Area Island problem in JavaScript. I'll write a clean, well-commented implementation that handles all edge cases.

```javascript
/**
 * Max Area Island - Find the maximum area of an island in a 2D binary grid
 * 
 * @param {number[][]} grid - 2D array where 1 represents land and 0 represents water
 * @return {number} - The area of the largest island
 * 
 * Time Complexity: O(m*n) where m is number of rows and n is number of columns
 * Space Complexity: O(m*n) in worst case for recursive call stack
 */

const maxAreaOfIsland = function(grid) {
    // Handle edge case of empty grid
    if (!grid || grid.length === 0) return 0;
    
    const rows = grid.length;
    const cols = grid[0].length;
    let maxArea = 0;
    
    // Helper function to check if a position is valid
    const isValid = (row, col) => {
        return row >= 0 && row < rows && col >= 0 && col < cols;
    };
    
    // DFS function to explore and count connected land cells
    const exploreIsland = (row, col) => {
        // Base cases: invalid position or water or visited cell
        if (!isValid(row, col) || grid[row][col] !== 1) {
            return 0;
        }
        
        // Mark current cell as visited by changing it to 2
        grid[row][col] = 2;
        
        // Explore all four directions and sum up the areas
        return 1 + // Current cell
            exploreIsland(row + 1, col) + // Down
            exploreIsland(row - 1, col) + // Up
            exploreIsland(row, col + 1) + // Right
            exploreIsland(row, col - 1);  // Left
    };
    
    // Iterate through each cell in the grid
    for (let row = 0; row < rows; row++) {
        for (let col = 0; col < cols; col++) {
            // If we find a land cell, explore the island
            if (grid[row][col] === 1) {
                maxArea = Math.max(maxArea, exploreIsland(row, col));
            }
        }
    }
    
    return maxArea;
};

// Example usage and test cases
const testCases = [
    // Test Case 1: Grid with multiple islands
    [
        [1,1,0,0,0],
        [1,1,0,0,0],
        [0,0,0,1,1],
        [0,0,0,1,1]
    ],
    
    // Test Case 2: Grid with no islands
    [
        [0,0,0],
        [0,0,0],
        [0,0,0]
    ],
    
    // Test Case 3: Grid with single cell island
    [
        [1]
    ],
    
    // Test Case 4: Empty grid
    []
];

// Run test cases
testCases.forEach((grid, index) => {
    console.log(`Test Case ${index + 1}:`);
    console.log('Input Grid:', grid);
    console.log('Max Area:', maxAreaOfIsland(grid));
    console.log('---');
});

module.exports = maxAreaOfIsland;
```

This implementation provides a solution to the Max Area Island problem with the following features:

1. **Main Function**: `maxAreaOfIsland` takes a 2D grid as input and returns the area of the largest island.

2. **Helper Functions**:
   - `isValid`: Checks if a given position is within grid boundaries
   - `exploreIsland`: DFS function to explore connected land cells and calculate area

3. **Key Features**:
   - Handles edge cases (empty grid, no islands)
   - Uses DFS to explore islands
   - Marks visited cells to avoid counting them multiple times
   - Includes test cases for verification

4. **Time and Space Complexity**:
   - Time: O(m*n) where m is rows and n is columns
   - Space: O(m*n) worst case for recursive call stack

5. **Test Cases**:
   - Multiple islands
   - No islands
   - Single cell island
   - Empty grid

The code follows JavaScript best practices and includes clear comments for better understanding. It's structured as a complete, runnable file that can be imported and used in other parts of the project.